### ISSUE TITLE
Option for light(er)weight packaging

### ISSUE THREAD

**Issue by fastnsilver** [spring-projects/spring-boot#1813]
First of all Spring Boot is awesome.  What might make it even more awesome?  Have the Maven plugin package a .jar without all the dependencies; resolving them at first run.  See https://github.com/puniverse/capsule.  Perhaps Spring Boot could provide a flag for doing such?

**Comment by olivergierke** [comment]
The actual problem caused with fat-JAR deployments is that — especially if the ned of you CI pipeline installs the artifact into a repository — you basically accumulate a lot of artifacts that have other artifacts wrapped in them. Other artifacts that are actually also available in the repository.

What if the resolution process was something to opt in via e.g. a command line flag and you could express that you might want to run the resolution against a well known local folder (e.g. `.m2/repository` only)? /cc @ewolff

**Comment by ewolff** [comment]
In addition it would be great if such a `.m2/repository` could be created by Spring Boot.

**Comment by olivergierke** [comment]
I guess we' have to use means to invoke Maven (or `$buildSystem` in general) for the artifact resolution. Manually replacing artifacts in an already existing repository is probably a tricky thing, esp. considering that multiple of those resolutions could run in parallel.

I know that the actual list of dependencies could be precomputed so that the calculation doesn't have to be done at resolution time but even downloading manually those artifacts might cause issues with the artifact metadata (or we'd have to replicate the update of that metadata, which is even worse).

**Comment by ewolff** [comment]
Maybe I wasn't clear. I imagine `mvn spring-boot:package-dependencies` would give me all dependencies needed in a single `repository.zip` that I could unpack on a machine to prepare it for the application.

When I start the application `java -jar myjar.jar` it should somehow use them and download additional artifacts if needed. That way I can prepare a Docker base image that already includes the dependencies.

**Comment by olivergierke** [comment]
Where'd you store the packaged ZIP then? Wouldn't (that have to) be deployed into the repository, too?

**Comment by ewolff** [comment]
I'd extract it into the Docker base image.

**Comment by dsyer** [comment]
I had a look at this and threw together something that works. It could be tidied up a bit, but it seems like a useful direction: https://github.com/scratches/launcher.

**Comment by wilkinsona** [comment]
If we're going to do something with lighter weight packaging, @dsyer's thin launcher looks to be the way to go. Before we can consider making it part of Boot we need to get some feedback on it. So, if you're interested in this functionality, please check out [what Dave has done](https://github.com/dsyer/spring-boot-thin-launcher) and let us know how well it works for you.

**Comment by olivergierke** [comment]
I've just added the thin launcher to a branch in Spring Restbucks. Very easy setup, works just fine. However, I couldn't get any output about what it's doing using the advertised command line flags and filed [a ticket](https://github.com/dsyer/spring-boot-thin-launcher/issues/14).

**Comment by fastnsilver** [comment]
Really nice work @dsyer! You considered and solved for quite a few more use cases than I had originally thought of. I will definitely recommend this packaging approach going forward. Seems like it should be baked into Spring Boot Maven Plugin rather than as a separate dependency. But I certainly understand why.

I like that you can target a Maven repo with `thin.root`.  One thought occurred to me... and that's if a `settings.xml` in that location refers to an artifact management repository, like Nexus or Artifactory (that is responsible for resolving "in-house/private" artifacts).  If you happen to  deploy an artifact (e.g., executable .jar), that has dependencies on "in-house" artifacts, to a target deployment environment (e.g., VM or container), some care should be taken. I'm thinking of a scenario where e.g., Artifactory resides in a separate (potentially unreachable) network than where the artifact will be running. In that case some kind of mirroring or proxying must be setup in the target environment to minimize exposing any "ip". << Definitely outside the concern of this implementation.

That said I see you can enable `thin.dryrun` so if you package your .jar in a container you could run this first when packaging.  So, the result is that you have all the dependencies baked into your image anyway, just in a Maven repo, and not in your "thin jar".  Looks like you are doing something similar in the fork of `java-buildpack` for Cloud Forundry.

**Comment by dsyer** [comment]
@fastnsilver I believe your `settings.xml` should work out of the box (we use the maven libraries to resolve dependencies). Try it and post a new issue in the thin-launcher repo if not.

**Comment by philwebb** [comment]
Closing this one in favor of the [thin jar launcher](https://github.com/dsyer/spring-boot-thin-launcher)

**Comment by jameshugit** [comment]
thin plugin is great, but, most of my projects are multi-module ones, I don't want to deploy the project modules to a private maven repository, is there any way I can pack the project-module jars into the thin jar?

**Comment by philwebb** [comment]
@jameshugit You'd have to raise a request at https://github.com/spring-projects-experimental/spring-boot-thin-launcher/issues

**Comment by dsyer** [comment]
Like https://github.com/spring-projects-experimental/spring-boot-thin-launcher/issues/78 or https://github.com/spring-projects-experimental/spring-boot-thin-launcher/issues/15 (ie no need for a new issue there).

**Comment by Kylin824** [comment]
> thin plugin is great, but, most of my projects are multi-module ones, I don't want to deploy the project modules to a private maven repository, is there any way I can pack the project-module jars into the thin jar?

i have the same question
